### PR TITLE
Refactor Pathway class

### PR DESCRIPTION
**Pull request by stefsmeets**
This PR refactors the Pathway class. I endend up tackling a bunch of cascading issues. It changes the following:

- Add `.dims` attribute to `Pathway`. This simplifies a bunch of site transformations that before depended on passing a volume
- Add method to calculate `Pathway` total length
- Move `frac_sites` / `wrap` to methods that return sites instead of in-place updates
- `Pathway.path_over_structure` returns list of `PeriodicSite` which simplifies return statement and gives more options to user
- Move some plots to methods of Pathway
- Streamlines creation of Pathways object from Volume
- Path plots no longer needs Volume because Pathway now has dims
- Rename path functions to hopefully make them a bit more consistent

### Todo

- [x] https://github.com/GEMDAT-repos/gemdat-notebooks/pull/11

### CODE REVIEW

**Comment by stefsmeets** [comment]
Hi @SCiarella could you review this pr please? It may seem like a bit much, but I think it is easiest to start with the Pathway refactor, most of the other changes follow from there. And then Volume next. Changes to both classes enabled a bunch of other small changes and corrections that I included (e.g. plots).

File: src/gemdat/plots/matplotlib/_paths.py [hunk 1]
```diff
@@ -4,72 +4,68 @@
...
-    volume: Volume,
+    other_paths: list[Pathway] | None = None,
 ) -> plt.Figure:
     """Plot energy along specified path.
 
     Parameters
     ----------
-    paths : Pathway | list[Pathway]
+    path : Pathway | list[Pathway]
         Pathway object containing the energy along the path, or list of Pathways
```
Comment (SCiarella) [inline src/gemdat/plots/matplotlib/_paths.py:20]:
```suggestion
    path : Pathway 
        Pathway object containing the energy along the path
```

File: src/gemdat/volume.py [hunk 2]
```diff
@@ -358,6 +377,71 @@ def plot_3d(self, **kwargs):
         return plots.plot_3d(volume=self, **kwargs)
 
 
+class FreeEnergyVolume(Volume):
```
Comment (SCiarella) [inline src/gemdat/volume.py:380]: I like this solution :+1:

**Review by SCiarella — APPROVED** [review]
Looks excellent! :rocket: 

My only comment is that I am not convinced that `wrapped_sites` should be optional. I would make it mandatory, or at least default on `True`.

File: src/gemdat/path.py [hunk 3]
```diff
@@ -44,42 +55,77 @@ def total_energy(self):
...
+        return [(x % xdim, y % xdim, z % xdim) for x, y, z in self.sites]
+
+    def frac_sites(self, wrapped: bool = False) -> np.ndarray:
+        """Return fractional sites.
+
+        Parameters
+        ----------
+        wrapped : bool
+            If True, wrap coordinates to bounding box using
+            `.wrapped_sites()`.
```
[the remark this answers is not in the document]
Comment (stefsmeets) [reply src/gemdat/path.py:99]: The only reason is that the original code had non-wrapped sites by default and required calling the `.wrap()` function. I didn't see any code that used the unwrapped sites, so I am happy to change this.

File: src/gemdat/path.py [hunk 4]
```diff
@@ -473,7 +529,7 @@ def find_best_perc_path(F: Volume,
             best_path = path
 
     if best_path:
-        # Before returning, wrap the path in the original volume
-        best_path.wrap(F.dims)
```
[the remark this answers is not in the document]
Comment (stefsmeets) [reply src/gemdat/path.py:477]: The tricky bit is that the Pathway object is created inside functions (`optimal_path`, `optimal_n_paths`) that have no access to the `dims` attribute required for wrapping. I'm not sure of a good way to solve it for now.

**Pull request merged**